### PR TITLE
🐛♻️ Fix Save all behavior and Reorganize Flaw & Affect App-State

### PR DESCRIPTION
**Pull request by superbuggy**
# OSIDB-3664 Fix Save All Affect Section
# OSIDB-3666 Refactor app state organization
# OSIDB-3700 Fix undesirable sorting of edited affects

## Checklist:

- [x] Commits consolidated
- [x] Changelog updated
- [x] Test cases added/updated
- [x] Jira ticket updated

## Summary:

Fixes the save all button behavior with an extensive refactoring of 

## Changes:

- Fix save all button reverting affects to their original state.
- Fixes an unreported CVSS change tracking bug.
- Merges useAffectSelections and affect editing functions into a single store.
  - Editing functions were moved out of `FlawAffects` and into their own dedicated store module (`AffectsEditingStore`).
  - Selection logic combined with store logic to avoid modules with circular dependencies
    - Uses separate modules internal to the store to maintain organization
    - Uses dependency injection to achieve loosest coupling
  - Separate Flaw data and Flaw Model logic so that dependent stores and composables can use the flaw data without importing unused Flaw Model functionalities.

## Considerations:

The small bug necessitated a fix that presented an opportunity to either accumulate more technical debt, or to pay it off. Accumulated more technical debt would have increased the number of sources of truth for affects data, and posed a significant risk of data inconsistency in the application. This posed an added risk of dramatically increasing the cost of future development work around affects-related features.

Closes OSIDB-3664. Closes OSIDB-3666. Closes OSIDB-3700.

### CODE REVIEW

File: src/composables/usePaginationWithSettings.ts [hunk 1]
```diff
@@ -46,9 +46,16 @@ export function usePaginationWithSettings(itemsToPaginate: Ref<any[]>, options:
...
     const start = (currentPage.value - 1) * settings.value[setting];
     const end = start + settings.value[setting];
-    return itemsToPaginate.value.slice(start, end);
+    // return itemsToPaginate.value.slice(start, end);
+    return itemsToPaginate.value.filter((_, index) => start <= index && index < end);
+    // },
+    // set(newValue: any) {
+    //   console.log('set', newValue);
+    //   itemsToPaginate.value = newValue;
+    // },
```
Comment (C-Valen) [inline src/composables/usePaginationWithSettings.ts:58]: Why is this change needed?

File: src/components/FlawAffects/useEditableAffects.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import { matcherForAffect } from '@/utils/helpers';
+
+import { useAffectSelections } from './useAffectSelections';
+
+const { initialAffects } = useFlawAffectsModel();
+
+const { isAffectSelected, toggleAffectSelection } = useAffectSelections();
+
+export const affectsBeingEdited = ref<ZodAffectType[]>([]);
+export const affectValuesPriorEdit = ref<ZodAffectType[]>([]);
```
Comment (MrMarble) [inline src/components/FlawAffects/useEditableAffects.ts:18]: Since this refactor introduces global state, maybe it makes sense to move all this logic to a proper `pinia` store, so the components don't directly mutate this refs, it should also help debugging with the VueDevTools.

File: src/components/FlawAffects/__tests__/FlawAffects.spec.ts [hunk 3]
```diff
@@ -1,67 +1,120 @@
...
 import sampleTrackersQueryResult from '@/components/__tests__/__fixtures__/sampleTrackersQueryResult.json';
-import { osimEmptyFlawTest, osimFullFlawTest } from '@/components/__tests__/test-suite-helpers';
 
+import { useFlaw } from '@/composables/useFlaw';
+// import { useFlawModel } from '@/composables/useFlawModel';
+import { useFlawAffectsModel } from '@/composables/useFlawAffectsModel';
+
+// import sampleFlawFull from '@/__tests__/__fixtures__/sampleFlawFull.json';
+import { useAffectsEditingStore } from '@/stores/AffectsEditingStore';
+import { mountWithConfig, withSetup } from '@/__tests__/helpers';
```
Comment (C-Valen) [inline src/components/FlawAffects/__tests__/FlawAffects.spec.ts]: May we remove commented imports?

File: src/components/FlawAffects/__tests__/FlawAffects.spec.ts [hunk 4]
```diff
@@ -1,67 +1,120 @@
...
-  props: {
-    embargoed: flaw.embargoed,
-    flaw,
-    relatedFlaws: [flaw],
-    error: [],
-  },
-});
+let pinia: ReturnType<typeof createPinia>;
+
+async function useMocks(flaw: ZodFlawType) {
```
Comment (C-Valen) [inline src/components/FlawAffects/__tests__/FlawAffects.spec.ts:29]: Is this reusable? May we want to have it in a external file (e.g. test helpers)?

File: src/composables/useFlawAffectsModel.ts [hunk 5]
```diff
@@ -269,3 +289,7 @@ function areAffectsEqual(affect: ZodAffectType, otherAffect?: ZodAffectType, sho
   const excludeFields = pickBy((_, key) => !fieldsToExclude.includes(key));
   return equals(excludeFields(affect), excludeFields(otherAffect));
 }
+
+function areAffectsCvssEqual(affectCvss: ZodAffectCVSSType, otherAffectCvss: ZodAffectCVSSType) {
+  return equals(affectCvss, otherAffectCvss);
+}
```
Comment (C-Valen) [inline src/composables/useFlawAffectsModel.ts:295]: Is this abstraction providing any advantage than just using `equals` where needed?

**Review by C-Valen — COMMENTED** [review]
Minor notes/concerns.
As a general comment for the future I'd aim for separating refactor in individual PRs. If a bug fix is requiring or taking advantage from a refactor IMO is better to make a previous ticket + PR for easier reviewing and more organized code changes.

**Review by C-Valen — APPROVED** [review]

**Review by MrMarble — DISMISSED** [review]

**Review by C-Valen — APPROVED** [review]

**Review by MrMarble — APPROVED** [review]
